src/core/vmm.d

```diff
@@ -0,0 +1,489 @@
+    return ptr ? ptr[0..size] : null;
+}
+
+/**
+    Commit a set of pages that form the specified $(D range) of virtual memory.
+    Only after applying this call is any memory obtained via $(LREF reserve)
+    guaranteed to be accessible.
+
+    Note: This function and $(LREF reserve)
+    set read/write permissions but not execute.
+    The reason is that some OS' would set permissions on reserve
+    while others on commit. Also Posix makes the commit stage almost transparent.
+    Use $(LREF protect) to override permissions as required.
+*/
+bool commit(void[] range)
+{
+    return osCommit(range.ptr, range.length);
+}
+
+/**
+    Resets a set of pages that form the specified $(D range) of virtual memory.
+    This instructs an OS that data in the pages is no loger of interest to the application,
+    so it's free to reuse any of page involved while keeping mapping intact.
+    Subsequent access will result in either reading a prior data or a fresh page.
+
+*/
+bool reset(void[] range)
+{
+    return osReset(range.ptr, range.length);
+}
+
+
+/**
+    Decommit a set of pages that form the specified $(D range) of virtual memory.
+    While system retains the address space range, accessing memory inside of it
+    is illegal. The exact consequences of such illegal access are platform
+    dependent.
+
+*/
+bool decommit(void[] range)
+{
+    return osDecommit(range.ptr, range.length);
+}
+
+/**
+    Decommit and releases pages that were previously allocated and/or reserved.
+    $(D range) must contain exactly the same address range as the one
+    returned by $(LREF allocate) or $(LREF reserve).
+
+    It doesn't matter what the state of each page was (reserved vs commited),
+    the OS decommits pages as needed.
+*/
+bool free(void[] range)
+{
+    return osUnmap(range.ptr, range.length);
+}
+
+/**
+    Set permissions on the set of pages that form $(D range).
+*/
+bool protect(void[] range, MemoryAccess access)
+{
+    return osProtect(range.ptr, range.length, access);
+}
+
+/**
+    Lock pages that form $(D range) to prevent their paging out.
+*/
+bool lock(void[] range)
+{
+    return osLockPages(range.ptr, range.length);
+}
+
+/**
+    Undo the effect of locking (with $(LREF lock)) pages that form $(D range).
+*/
+bool unlock(void[] range)
+{
+    return osUnlockPages(range.ptr, range.length);
+}
+
+/**
+    Round up a pointer or integral $(D arg)
+    to the nearest multiple of platform's page size.
+*/
+T roundUpToPage(T)(T arg)
+    if (is(T : void*) || is(T : size_t))
+{
+    return cast(T)((cast(size_t)arg + pageSize-1) & ~(pageSize-1));
+}
+
+/**
+    Round down a pointer or integral $(D arg)
+    to the nearest multiple of platform's page size.
+*/
+T roundDownToPage(T)(T arg)
+    if (is(T : void*) || is(T : size_t))
+{
+    return cast(T)(cast(size_t)arg & ~(pageSize-1));
+}
+
+//this is not even nearly covering all cases
+//some things are hard to check (how to check that segmentation fault is generated?)
```

Comment:
FWIW, it's possible to test whether a page is readable or writable without causing a segmentation fault on Windows using `IsBadRead/WritePtr` and on UNIX [using a `read`/`write` syscall with a pipe](http://stackoverflow.com/questions/7134590/how-to-test-if-an-address-is-readable-in-linux-userspace-app). Perhaps these can be worked into primitives for this module, as well.